cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java

```diff
@@ -78,38 +89,63 @@
   private final long pollTimeMillis;
   private final long gracefulShutdownMillis;
   private final List<MonitorMessage> lastProgramStateMessages;
+  private final DatasetFramework dsFramework;
+  private final MultiThreadMessagingContext messagingContext;
+  private final Transactional transactional;
+  private final RetryStrategy retryStrategy;
+  private AppMetadataStore store;
   private volatile Thread runThread;
   private volatile boolean stopped;
+  private volatile boolean killed;
   private long programFinishTime;
 
   public RuntimeMonitor(ProgramRunId programRunId, CConfiguration cConf,
-                        MessagingService messagingService, ClientConfig clientConfig) {
+                        MessagingService messagingService, ClientConfig clientConfig,
+                        DatasetFramework datasetFramework, TransactionSystemClient txClient) {
     this.programRunId = programRunId;
     this.cConf = cConf;
-    this.messagePublisher = new MultiThreadMessagingContext(messagingService).getMessagePublisher();
     this.clientConfig = clientConfig;
     this.restClient = new RESTClient(clientConfig);
     this.limit = cConf.getInt(Constants.RuntimeMonitor.BATCH_LIMIT);
     this.pollTimeMillis = cConf.getLong(Constants.RuntimeMonitor.POLL_TIME_MS);
     this.gracefulShutdownMillis = cConf.getLong(Constants.RuntimeMonitor.GRACEFUL_SHUTDOWN_MS);
     this.topicsToRequest = new HashMap<>();
     this.requestKeyToLocalTopic = new HashMap<>();
+    this.dsFramework = datasetFramework;
+    this.messagingContext = new MultiThreadMessagingContext(messagingService);
+    this.transactional = Transactions.createTransactionalWithRetry(
+      Transactions.createTransactional(new MultiThreadDatasetCache(
+        new SystemDatasetInstantiator(datasetFramework), new TransactionSystemClientAdapter(txClient),
+        NamespaceId.SYSTEM, Collections.emptyMap(), null, null, messagingContext)),
+      org.apache.tephra.RetryStrategies.retryOnConflict(20, 100)
+    );
     this.programFinishTime = -1L;
     this.lastProgramStateMessages = new ArrayList<>();
+    this.retryStrategy = RetryStrategies.timeLimit(cConf.getLong(Constants.RuntimeMonitor.RETRY_TIMEOUT_MS),
+                                                   TimeUnit.MILLISECONDS,
+                                                   RetryStrategies.fixDelay(1, TimeUnit.SECONDS));
   }
 
   @Override
   protected void startUp() throws Exception {
     LOG.debug("Starting runtime monitor for program run {}.", programRunId);
-    // TODO initialize from offset table for a given programId
-    for (String topicConfig : getTopicConfigs()) {
-      topicsToRequest.put(topicConfig, new MonitorConsumeRequest(null, limit));
+
+    Set<String> topicConfigs = getTopicConfigs();
+    for (String topicConfig : topicConfigs) {
       requestKeyToLocalTopic.put(topicConfig, getTopic(topicConfig));
     }
+
+    Retries.runWithRetries(() -> Transactionals.execute(transactional, context -> {
```

Comment:
Why need to do this in the `startUp`? You can do it as part of the run loop such that all failures are handled in the same way (basically retry in the next iteration). Something like:

```java
void run() {
  while (...) {
    try {
      if (topicsToRequest.isEmpty()) {
        initTopics(topicsToRequest);
      }
      // Inside a transaction do {
        // Fetching and republishing logic
        // Update persisted states
      // }
      // No exception, hence update the in memory states
      sleepTime = pollTimeMillis;
    } catch (...) {
      // Use Retry to compute the sleep time
      sleepTime = ...;
    }

    if (!stopped && !killed) {
      TimeUnit.MILLISECONDS.sleep(sleepTime);
    }
  }
}
```